ubuntu_image/classic_builder.py

```diff
@@ -0,0 +1,575 @@
+            args=self.args,
+            cloud_init=self.cloud_init,
+            done=self.done,
+            exitcode=self.exitcode,
+            gadget=self.gadget,
+            output=self.output,
+            output_dir=self.output_dir,
+            part_images=self.part_images,
+            rootfs=self.rootfs,
+            rootfs_size=self.rootfs_size,
+            unpackdir=self.unpackdir,
+            volumedir=self.volumedir,
+            gadget_tree=self.gadget_tree,
+            hookdirs=self.hookdirs,
+            )
+        return state
+
+    def __setstate__(self, state):
+        super().__setstate__(state)
+        self.args = state['args']
+        self.cloud_init = state['cloud_init']
+        self.done = state['done']
+        self.exitcode = state['exitcode']
+        self.gadget = state['gadget']
+        self.output = state['output']
+        self.output_dir = state['output_dir']
+        self.part_images = state['part_images']
+        self.rootfs = state['rootfs']
+        self.rootfs_size = state['rootfs_size']
+        self.gadget_tree = state['gadget_tree']
+        self.unpackdir = state['unpackdir']
+        self.volumedir = state['volumedir']
+        self.hookdirs = state['hookdirs']
+        # Restore the hook manager along with the state.
+        self.hook_manager = HookManager(self.hookdirs)
+
+    def _log_exception(self, name):
+        # Only log the exception if we're in debug mode.
+        if self.args.debug:
+            super()._log_exception(name)
+
+    def make_temporary_directories(self):
+        self.rootfs = os.path.join(self.workdir, 'root')
+        self.unpackdir = os.path.join(self.workdir, 'unpack')
+        self.volumedir = os.path.join(self.workdir, 'volumes')
+
+        os.makedirs(self.rootfs)
+
+        self._next.append(self.prepare_gadget_tree)
+
+    def prepare_gadget_tree(self):
+        try:
+            gadget_dst = os.path.join(self.unpackdir, 'gadget_repo')
+            if urlparse(self.gadget_tree).scheme != "":
+                git.Repo.clone_from(self.gadget_tree, gadget_dst,
+                                    progress=GitProgress())
+            else:
+                shutil.copytree(self.gadget_tree, gadget_dst)
+            # generate unpacked gadget snap
+            with save_cwd():
+                os.chdir(gadget_dst)
+                run(['snapcraft', 'prime'], env=os.environ)
+            shutil.copytree(os.path.join(gadget_dst, 'prime'),
+                            os.path.join(self.unpackdir, 'gadget'))
+
+        except CalledProcessError:
+            if self.args.debug:
+                _logger.exception('Full debug traceback follows')
+            self.exitcode = 1
+            # Stop the state machine right here by not appending a next step.
+        else:
+            self._next.append(self.prepare_image)
+
+    def prepare_image(self):
+        # It's required to run ubuntu-image as root to build classic image.
+        check_root_privilege()
```

Comment:
Thinking about it now, a question - since I didn't run the branch locally yet for a full image build. Are the root privilages required only for running live-build or do we need those also for the file copying to the rootfs target directory? Since if it's only the former, then I think it would be more user-sensible if, when someone runs ubuntu-image not as root, he gets a prompt from sudo to type in the password. That's what happens when you run ubuntu-image as a normal user on xenial when building snap images. Root is required there as there's a 'sudo cp' operation (for later series it's not needed).

Reply:
All files in rootfs generated by live-build are owned by root. We manually moved those files to target rootfs directory in `populate_rootfs_contents` by using python-native method. Meanwhile,  considering the rootfs might be changed at hooks state. Root privilege is required for all those states.  
We definitely can run those state through shell with `sudo` prepended alternatively. But I've been reluctant to do this because we need to keep the same code pattern(python-native) in both core and classic builder. I tend to keep the current implementation. 
How is your option?

RE: pep8/pylint cleanand the missing gadget_tree issue
I'll fix those in the next commit.

Reply:
Ok, regarding this - since I would really prefer if we removed the need of running the whole ubuntu-image as root. Root privilages are basically only required for running lb, so we should only need them there. This way we can potentially test more pieces.

A simple, not-too-pretty solution would be to get rid of check_root_privilege() and instead modify populate_rootfs_contents() slightly. Let me elaborate more on the line below.